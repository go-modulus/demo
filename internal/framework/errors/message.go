package errors

import (
	"errors"

	"golang.org/x/text/message"
)

func WithMessage(err error, message string) error {
	if err == nil {
		return nil
	}
	return withMessage{
		cause:   err,
		message: message,
	}
}

type withMessage struct {
	cause   error
	message string
}

func (w withMessage) Error() string   { return w.cause.Error() }
func (w withMessage) Cause() error    { return w.cause }
func (w withMessage) Message() string { return w.message }

// Unwrap provides compatibility for Go 1.13 error chains.
func (w withMessage) Unwrap() error { return w.cause }

func Message(t *message.Printer, err error) string {
	type withMessage interface {
		Message() string
	}
	var wm withMessage
	if errors.As(err, &wm) {
		return wm.Message()
	}

	type withMessagePrinter interface {
		Message(*message.Printer) string
	}
	var wmp withMessagePrinter
	if errors.As(err, &wmp) {
		return wmp.Message(t)
	}

	return "Internal Server Error"
}
