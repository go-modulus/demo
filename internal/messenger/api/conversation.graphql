extend type Query {
    conversation(id: Uuid!): ConversationResult!
    myConversations(first: Int! = 25, after: String): MyConversationsResult!
}

extend type Mutation {
    createOneToOneConversation(receiverId: Uuid!): CreateOneToOneConversationResult!
}

union ConversationResult = ConversationBox | ErrUnknownConversation | ErrInvalidInput | ErrUnauthorized

type ErrUnknownConversation implements Error {
    message: String!
}

union MyConversationsResult = ConversationList | ErrInvalidInput | ErrUnauthorized

union CreateOneToOneConversationResult = OneToOneConversation | ErrUnknownReceiver | ErrInvalidInput | ErrUnauthorized

type ErrUnknownReceiver implements Error {
    message: String!
}

union Conversation = OneToOneConversation

type ConversationBox {
    value: Conversation!
}

type OneToOneConversation {
    id: Uuid!
    draft: Draft @goField(forceResolver: true)
    lastMessage: Message @goField(forceResolver: true)
    createdAt: MilliTimestamp!
}

type ConversationEdge {
    cursor: String!
    node: Conversation!
}

type ConversationList {
    edges: [ConversationEdge]!
    pageInfo: PageInfo!
}

directive @goField(
    forceResolver: Boolean
    name: String
) on INPUT_FIELD_DEFINITION | FIELD_DEFINITION | INTERFACE