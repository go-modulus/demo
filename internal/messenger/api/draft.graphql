extend type Mutation {
    saveDraft(conversationId: Uuid!, messageId: Uuid, text: String): SaveDraftResult!
    removeDraft(conversationId: Uuid!): RemoveDraftResult!
}

union SaveDraftResult = Draft | ErrUnknownConversation | ErrUnknownMessage | ErrInvalidInput | ErrUnauthorized
union RemoveDraftResult = VoidBox | ErrUnknownConversation | ErrInvalidInput | ErrUnauthorized

type Draft {
    conversationId: Uuid!
    messageId: Uuid
    richText: RichText
}