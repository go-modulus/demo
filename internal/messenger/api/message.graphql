extend type Query {
    messages(conversationId: Uuid!, first: Int! = 25, after: String): MessagesResult!
}

extend type Mutation {
    createMessage(conversationId: Uuid!, text: String): CreateMessageResult!
    editMessage(messageId: Uuid!, text: String): EditMessageResult!
    deleteMessage(messageId: Uuid!): DeleteMessageResult!
}

union MessagesResult = MessageList | ErrUnknownConversation | ErrInvalidInput | ErrUnauthorized

union CreateMessageResult = TextMessage | ErrUnknownConversation | ErrTextTooLong | ErrInvalidInput | ErrUnauthorized
union EditMessageResult = TextMessage | ErrUnknownMessage | ErrTextTooLong | ErrInvalidInput | ErrUnauthorized
union DeleteMessageResult = VoidBox | ErrUnknownMessage | ErrInvalidInput | ErrUnauthorized

type ErrUnknownMessage implements Error {
    message: String!
}

type ErrTextTooLong implements Error {
    message: String!

    currentLength: Int!
    maxLength: Int!
}

union Message = TextMessage

type TextMessage {
    id: Uuid!
    conversationId: Uuid!
    richText: RichText
    updatedAt: MilliTimestamp!
    createdAt: MilliTimestamp!
}

type MessageEdge {
    cursor: String!
    node: Message!
}

type MessageList {
    edges: [MessageEdge]!
    pageInfo: PageInfo!
}

type RichText {
    text: String
    parts: [RichTextPart!]!
}

union RichTextPart = PlainRichText

type PlainRichText {
    text: String!
}