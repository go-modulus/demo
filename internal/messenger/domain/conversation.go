package domain

import (
	"context"
	"database/sql"
	"fmt"
	"github.com/gofrs/uuid"
	"time"
)

type ErrConversationNotFound struct {
	ConversationId uuid.UUID
}

func (e ErrConversationNotFound) Error() string {
	return fmt.Sprintf("conversation with id %s not found", e.ConversationId)
}

type Conversation struct {
	Id         uuid.UUID
	SenderId   uuid.UUID
	ReceiverId uuid.UUID
	UpdatedAt  time.Time
	CreatedAt  time.Time
}

func (c Conversation) IsParticipant(personId uuid.UUID) bool {
	return c.SenderId == personId || c.ReceiverId == personId
}

type ConversationRepository interface {
	Get(ctx context.Context, id uuid.UUID) (*Conversation, error)
	Paginate(ctx context.Context, viewer uuid.UUID, first int, after sql.NullString) (*Page[*Conversation], error)

	GetOrCreate(ctx context.Context, sender uuid.UUID, receiver uuid.UUID) (*Conversation, error)
}

type ConversationCreator struct {
	conversations ConversationRepository
}

func NewConversationCreator(conversations ConversationRepository) *ConversationCreator {
	return &ConversationCreator{conversations: conversations}
}

func (c *ConversationCreator) Create(ctx context.Context, sender uuid.UUID, receiver uuid.UUID) (*Conversation, error) {
	conversation, err := c.conversations.GetOrCreate(ctx, sender, receiver)
	if err != nil {
		return nil, err
	}

	return conversation, nil
}

type ConversationViewer struct {
	conversations ConversationRepository
	policy        *ConversationPolicy
}

func NewConversationViewer(conversations ConversationRepository, policy *ConversationPolicy) *ConversationViewer {
	return &ConversationViewer{conversations: conversations, policy: policy}
}

func (c *ConversationViewer) View(ctx context.Context, viewer uuid.UUID, id uuid.UUID) (*Conversation, error) {
	conversation, err := c.conversations.Get(ctx, id)
	if err != nil {
		return nil, err
	}

	if !c.policy.CanView(viewer, conversation) {
		return nil, ErrConversationNotFound{ConversationId: id}
	}

	return conversation, nil
}

func (c *ConversationViewer) Paginate(
	ctx context.Context,
	viewer uuid.UUID,
	first int,
	after sql.NullString,
) (*Page[*Conversation], error) {
	page, err := c.conversations.Paginate(ctx, viewer, first, after)
	if err != nil {
		return nil, err
	}

	return page, nil
}

type ConversationPolicy struct{}

func NewConversationPolicy() *ConversationPolicy {
	return &ConversationPolicy{}
}

func (c *ConversationPolicy) CanView(personId uuid.UUID, conversation *Conversation) bool {
	if conversation == nil {
		return false
	}

	return conversation.IsParticipant(personId)
}

func (c *ConversationPolicy) CanSendMessages(personId uuid.UUID, conversation *Conversation) bool {
	if conversation == nil {
		return false
	}

	return conversation.IsParticipant(personId)
}
