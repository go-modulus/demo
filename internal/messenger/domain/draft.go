package domain

import (
	"context"
	"github.com/gofrs/uuid"
)

type Draft struct {
	Id             uuid.UUID
	ConversationId uuid.UUID
	AuthorId       uuid.UUID
	MessageId      uuid.NullUUID
	Text           RichText
}

type DraftRepository interface {
	FindDrafts(
		ctx context.Context,
		authorId uuid.UUID,
		conversationIds []uuid.UUID,
	) (map[uuid.UUID]*Draft, error)
	FindOrCreateByConversationAndAuthor(
		ctx context.Context,
		conversationId uuid.UUID,
		authorId uuid.UUID,
	) (*Draft, error)
	Update(
		ctx context.Context,
		id uuid.UUID,
		updateFunc func(ctx context.Context, draft *Draft) (*Draft, error),
	) error
	RemoveByConversationAndAuthor(
		ctx context.Context,
		conversationId uuid.UUID,
		authorId uuid.UUID,
	) error
}

type SaveDraft struct {
	ConversationId uuid.UUID
	SenderId       uuid.UUID
	MessageId      uuid.NullUUID
	Text           string
}

type DraftSaver struct {
	conversations      ConversationRepository
	conversationPolicy *ConversationPolicy
	messages           MessageRepository
	messagePolicy      *MessagePolicy
	richTextParser     *RichTextParser
	drafts             DraftRepository
}

func NewDraftSaver(
	conversations ConversationRepository,
	conversationPolicy *ConversationPolicy,
	messages MessageRepository,
	messagePolicy *MessagePolicy,
	richTextParser *RichTextParser,
	drafts DraftRepository,
) *DraftSaver {
	return &DraftSaver{
		conversations:      conversations,
		conversationPolicy: conversationPolicy,
		messages:           messages,
		messagePolicy:      messagePolicy,
		richTextParser:     richTextParser,
		drafts:             drafts,
	}
}

func (s *DraftSaver) Save(ctx context.Context, dto SaveDraft) (*Draft, error) {
	conversation, err := s.conversations.Get(ctx, dto.ConversationId)
	if err != nil {
		return nil, err
	}
	if !s.conversationPolicy.CanView(dto.SenderId, conversation) {
		return nil, ErrConversationNotFound{ConversationId: dto.ConversationId}
	}

	var message *Message
	if dto.MessageId.Valid {
		message, err = s.messages.Get(ctx, dto.MessageId.UUID)
		if err != nil {
			return nil, err
		}
		if !s.messagePolicy.CanEdit(dto.SenderId, message) {
			return nil, ErrMessageNotFound{MessageId: dto.MessageId.UUID}
		}
		if message.ConversationId != conversation.Id {
			return nil, ErrMessageNotFound{MessageId: dto.MessageId.UUID}
		}
	}

	richText, err := s.richTextParser.Parse(dto.Text)
	if err != nil {
		return nil, err
	}

	draftToReturn, err := s.drafts.FindOrCreateByConversationAndAuthor(
		ctx,
		conversation.Id,
		dto.SenderId,
	)
	if err != nil {
		return nil, err
	}

	err = s.drafts.Update(
		ctx,
		draftToReturn.Id,
		func(ctx context.Context, draft *Draft) (*Draft, error) {
			if message != nil {
				draft.MessageId = uuid.NullUUID{UUID: message.Id, Valid: true}
			} else {
				draft.MessageId = uuid.NullUUID{}
			}

			draft.Text = richText
			draftToReturn = draft

			return draft, nil
		},
	)
	if err != nil {
		return nil, err
	}

	return draftToReturn, nil
}

type RemoveDraft struct {
	ConversationId uuid.UUID
	AuthorId       uuid.UUID
}

type DraftRemover struct {
	drafts DraftRepository
}

func NewDraftRemover(drafts DraftRepository) *DraftRemover {
	return &DraftRemover{drafts: drafts}
}

func (s *DraftRemover) Remove(ctx context.Context, dto RemoveDraft) error {
	return s.drafts.RemoveByConversationAndAuthor(ctx, dto.ConversationId, dto.AuthorId)
}
