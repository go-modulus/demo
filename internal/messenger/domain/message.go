package domain

import (
	"context"
	"database/sql"
	"fmt"
	"github.com/gofrs/uuid"
	"time"
)

type ErrMessageNotFound struct {
	MessageId uuid.UUID
}

func (e ErrMessageNotFound) Error() string {
	return fmt.Sprintf("message with id %s not found", e.MessageId)
}

var (
	MessageStatusUnknown = MessageStatus{""}
	MessageStatusNew     = MessageStatus{"confirmed"}
	MessageStatusDeleted = MessageStatus{"deleted"}
)

type MessageStatus struct {
	value string
}

func NewMessageStatus(value string) MessageStatus {
	switch value {
	case MessageStatusNew.value:
		return MessageStatusNew
	case MessageStatusDeleted.value:
		return MessageStatusDeleted
	}

	return MessageStatusUnknown
}

type MessageCreated struct {
	MessageId      uuid.UUID
	ConversationId uuid.UUID
	SenderId       uuid.UUID
	Text           RichText
	CreatedAt      time.Time
	BaseEvent
}

type Message struct {
	Id             uuid.UUID
	ConversationId uuid.UUID
	SenderId       uuid.UUID
	Text           RichText
	Status         MessageStatus
	UpdatedAt      time.Time
	CreatedAt      time.Time
	EventCollector
}

func (m *Message) MarkAsDeleted() {
	if m.Status == MessageStatusDeleted {
		return
	}

	m.Status = MessageStatusDeleted
	m.UpdatedAt = time.Now()
}

func (m *Message) IsSender(personId uuid.UUID) bool {
	return m.SenderId == personId
}

func NewMessage(conversationID uuid.UUID, senderID uuid.UUID, richText RichText) *Message {
	now := time.Now()

	message := &Message{
		Id:             uuid.Must(uuid.NewV6()),
		ConversationId: conversationID,
		SenderId:       senderID,
		Text:           richText,
		UpdatedAt:      now,
		CreatedAt:      now,
	}

	message.recordThat(&MessageCreated{
		MessageId:      message.Id,
		ConversationId: message.ConversationId,
		SenderId:       message.SenderId,
		Text:           message.Text,
		CreatedAt:      message.CreatedAt,
	})

	return message
}

type MessageRepository interface {
	Get(ctx context.Context, id uuid.UUID) (*Message, error)
	FindLastMessages(ctx context.Context, conversationIds []uuid.UUID) (map[uuid.UUID]*Message, error)
	Paginate(ctx context.Context, conversationId uuid.UUID, first int, after sql.NullString) (*Page[*Message], error)
	Update(
		ctx context.Context,
		id uuid.UUID,
		updateFunc func(ctx context.Context, draft *Message) (*Message, error),
	) error
	Add(ctx context.Context, message *Message) error
}

type CreateMessage struct {
	PerformerId    uuid.UUID
	ConversationId uuid.UUID
	Text           string
}

type MessageCreator struct {
	richTextParser *RichTextParser
	messages       MessageRepository
}

func NewMessageCreator(richTextParser *RichTextParser, messages MessageRepository) *MessageCreator {
	return &MessageCreator{richTextParser: richTextParser, messages: messages}
}

func (c *MessageCreator) Create(ctx context.Context, dto CreateMessage) (*Message, error) {
	richText, err := c.richTextParser.Parse(dto.Text)
	if err != nil {
		return nil, err
	}

	message := NewMessage(dto.ConversationId, dto.PerformerId, richText)

	if err := c.messages.Add(ctx, message); err != nil {
		return nil, err
	}

	return message, nil
}

type EditMessage struct {
	PerformerId uuid.UUID
	MessageId   uuid.UUID
	Text        string
}

type MessageEditor struct {
	richTextParser *RichTextParser
	messages       MessageRepository
}

func NewMessageEditor(richTextParser *RichTextParser, messages MessageRepository) *MessageEditor {
	return &MessageEditor{richTextParser: richTextParser, messages: messages}
}

func (c *MessageEditor) Edit(ctx context.Context, dto EditMessage) (*Message, error) {
	richText, err := c.richTextParser.Parse(dto.Text)
	if err != nil {
		return nil, err
	}

	var messageToReturn *Message
	err = c.messages.Update(
		ctx,
		dto.MessageId,
		func(ctx context.Context, message *Message) (*Message, error) {
			message.Text = richText
			message.UpdatedAt = time.Now()
			messageToReturn = message
			return message, nil
		},
	)

	if err != nil {
		return nil, err
	}

	return messageToReturn, nil
}

type RemoveMessage struct {
	PerformerId uuid.UUID
	MessageId   uuid.UUID
}

type MessageRemover struct {
	messages MessageRepository
	policy   *MessagePolicy
}

func NewMessageRemover(messages MessageRepository, policy *MessagePolicy) *MessageRemover {
	return &MessageRemover{messages: messages, policy: policy}
}

func (r *MessageRemover) Remove(ctx context.Context, dto RemoveMessage) error {
	return r.messages.Update(
		ctx,
		dto.MessageId,
		func(ctx context.Context, message *Message) (*Message, error) {
			if !r.policy.CanDelete(dto.PerformerId, message) {
				return nil, ErrMessageNotFound{MessageId: dto.MessageId}
			}

			message.MarkAsDeleted()

			return message, nil
		},
	)
}

type MessageViewer struct {
	conversations      ConversationRepository
	conversationPolicy *ConversationPolicy
	messages           MessageRepository
}

func NewMessageViewer(
	conversations ConversationRepository,
	conversationPolicy *ConversationPolicy,
	messages MessageRepository,
) *MessageViewer {
	return &MessageViewer{
		conversations:      conversations,
		conversationPolicy: conversationPolicy,
		messages:           messages,
	}
}

func (c *MessageViewer) Paginate(
	ctx context.Context,
	performerId uuid.UUID,
	conversationId uuid.UUID,
	first int,
	after sql.NullString,
) (*Page[*Message], error) {
	conversation, err := c.conversations.Get(ctx, conversationId)
	if err != nil {
		return nil, err
	}
	if !c.conversationPolicy.CanView(performerId, conversation) {
		return nil, ErrConversationNotFound{ConversationId: conversationId}
	}

	page, err := c.messages.Paginate(ctx, conversationId, first, after)
	if err != nil {
		return nil, err
	}

	return page, nil
}

type MessagePolicy struct{}

func NewMessagePolicy() *MessagePolicy {
	return &MessagePolicy{}
}

func (c *MessagePolicy) CanEdit(personId uuid.UUID, message *Message) bool {
	if message == nil {
		return false
	}

	return message.IsSender(personId)
}

func (c *MessagePolicy) CanDelete(personId uuid.UUID, message *Message) bool {
	if message == nil {
		return false
	}

	return message.IsSender(personId)
}
