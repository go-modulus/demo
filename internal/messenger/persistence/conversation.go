package persistence

import (
	"context"
	"database/sql"
	"demo/internal/errors"
	"demo/internal/messenger/domain"
	"encoding/base64"
	"encoding/json"
	"github.com/gofrs/uuid"
	"time"
)

var _ domain.ConversationRepository = &ConversationRepository{}

type ConversationRepository struct {
	queries *Queries
}

func NewConversationRepository(queries *Queries) *ConversationRepository {
	return &ConversationRepository{queries: queries}
}

func (c *ConversationRepository) Get(ctx context.Context, id uuid.UUID) (*domain.Conversation, error) {
	conversation, err := c.queries.GetConversation(ctx, id)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, domain.ErrConversationNotFound{ConversationId: id}
	}
	if err != nil {
		return nil, err
	}

	return &domain.Conversation{
		Id:         conversation.ID,
		SenderId:   conversation.SenderID,
		ReceiverId: conversation.ReceiverID,
		UpdatedAt:  conversation.UpdatedAt,
		CreatedAt:  conversation.CreatedAt,
	}, nil
}

func (c *ConversationRepository) Paginate(
	ctx context.Context,
	viewer uuid.UUID,
	first int,
	after sql.NullString,
) (*domain.Page[*domain.Conversation], error) {
	params := PaginateMyConversationsParams{
		ViewerID: viewer,
		First:    int32(first) + 1,
	}

	if after.Valid {
		cursor, err := c.decodeCursor(after.String)
		if err != nil {
			return nil, err
		}

		if rawUpdatedAt, ok := cursor["updatedAt"]; ok {
			params.AfterUpdatedAt = sql.NullTime{
				Time:  time.UnixMicro(int64(rawUpdatedAt.(float64))),
				Valid: true,
			}

			if rawId, ok := cursor["id"]; ok {
				id, err := uuid.FromString(rawId.(string))
				if err != nil {
					return nil, err
				}

				params.AfterID = uuid.NullUUID{
					UUID:  id,
					Valid: true,
				}
			}
		}
	}

	conversations, err := c.queries.PaginateMyConversations(ctx, params)
	if err != nil {
		return nil, err
	}

	edges := make([]domain.Edge[*domain.Conversation], 0, len(conversations))
	for i, conversation := range conversations {
		if i == first {
			break
		}

		cursor, err := c.encodeCursor(map[string]any{
			"updatedAt": conversation.UpdatedAt.UnixMicro(),
			"id":        conversation.ID.String(),
		})
		if err != nil {
			return nil, err
		}

		edges = append(edges, domain.Edge[*domain.Conversation]{
			Cursor: cursor,
			Node:   c.mapToDomain(conversation),
		})
	}

	var startCursor, endCursor *string
	if len(edges) > 0 {
		if len(edges) > 0 {
			startCursor = &edges[0].Cursor
			endCursor = &edges[len(edges)-1].Cursor
		}
	}

	return &domain.Page[*domain.Conversation]{
		Edges:           edges,
		HasNextPage:     len(conversations) > len(edges),
		HasPreviousPage: after.Valid,
		StartCursor:     startCursor,
		EndCursor:       endCursor,
	}, nil
}

func (c *ConversationRepository) GetOrCreate(ctx context.Context, sender uuid.UUID, receiver uuid.UUID) (*domain.Conversation, error) {
	conversation, err := c.queries.CreateOrGetConversation(ctx, CreateOrGetConversationParams{
		ID:         uuid.Must(uuid.NewV6()),
		SenderID:   sender,
		ReceiverID: receiver,
		UpdatedAt:  time.Now(),
		CreatedAt:  time.Now(),
	})

	if err != nil {
		return nil, err
	}

	return &domain.Conversation{
		Id:         conversation.ID,
		SenderId:   conversation.SenderID,
		ReceiverId: conversation.ReceiverID,
		UpdatedAt:  conversation.UpdatedAt,
		CreatedAt:  conversation.CreatedAt,
	}, nil
}

func (c *ConversationRepository) encodeCursor(cursor map[string]any) (string, error) {
	rawJson, err := json.Marshal(cursor)
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(rawJson), nil
}

func (c *ConversationRepository) decodeCursor(rawCursor string) (map[string]any, error) {
	rawJson, err := base64.StdEncoding.DecodeString(rawCursor)
	if err != nil {
		return nil, err
	}

	cursor := make(map[string]any)
	err = json.Unmarshal(rawJson, &cursor)
	if err != nil {
		return nil, err
	}

	return cursor, nil
}

func (c *ConversationRepository) mapToDomain(conversation MessengerConversation) *domain.Conversation {
	return &domain.Conversation{
		Id:         conversation.ID,
		SenderId:   conversation.SenderID,
		ReceiverId: conversation.ReceiverID,
		UpdatedAt:  conversation.UpdatedAt,
		CreatedAt:  conversation.CreatedAt,
	}
}
