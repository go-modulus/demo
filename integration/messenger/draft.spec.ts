import {persons} from "../utils";
import {getOrCreatePool, performScheduledDeletions, scheduleDeletion, sql} from "../db";
import {createConversation} from "./conversation/create";
import {revertSaveDraft, saveDraft, sendSaveDraftMutation} from "./draft/save";
import {sendRemoveDraftMutation} from "./draft/remove";
import {createMessage} from "./message/create";

const clearDb = async () => {
    const pool = await getOrCreatePool();

    await pool.connect(async (connection) => {
        const personIds = Object.values(persons)

        await connection.query(
            sql.typeAlias('void')`
                DELETE
                FROM messenger.draft
                WHERE author_id IN (${sql.join(personIds, sql.fragment`, `)})
            `
        )
    })
}

beforeAll(clearDb)
afterEach(performScheduledDeletions)
afterAll(async () => {
    const pool = await getOrCreatePool();
    if (pool.getPoolState().ended) {
        return;
    }

    await pool.end();
})

test("save draft", async () => {
    const conversationId = await createConversation(persons.alice, persons.bob);
    const messageId = await createMessage(persons.alice, conversationId, 'Hi')

    const res = await sendSaveDraftMutation(persons.alice, conversationId, messageId, 'Hi!');
    scheduleDeletion(() => revertSaveDraft(persons.alice, conversationId))

    expect(res.error).toBeFalsy();
    expect(res.body).toMatchSnapshot({
        data: {
            saveDraft: {
                conversationId: expect.any(String),
                messageId: expect.any(String),
            },
        },
    });
    expect(res.body).toMatchObject({
        data: {
            saveDraft: {
                conversationId,
                messageId,
            },
        },
    })
});

test("remove draft", async () => {
    const conversationId = await createConversation(persons.alice, persons.bob);
    await saveDraft(persons.alice, conversationId, null, 'Hi!');

    const res = await sendRemoveDraftMutation(persons.alice, conversationId);

    expect(res.error).toBeFalsy();
    expect(res.body).toMatchSnapshot();
});
