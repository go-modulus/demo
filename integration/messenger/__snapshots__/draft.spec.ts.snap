// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remove draft 1`] = `
{
  "data": {
    "removeDraft": null,
  },
}
`;

exports[`save draft 1`] = `
{
  "data": {
    "saveDraft": {
      "conversationId": Any<String>,
      "messageId": Any<String>,
      "richText": {
        "parts": [
          {
            "__typename": "PlainRichText",
            "text": "Hi!",
          },
        ],
        "text": "Hi!",
      },
    },
  },
}
`;
